### ISSUE TITLE
ASTs for incomplete code

### ISSUE THREAD

**Issue by akhileshgotmare** [tech-srl/slm-code-generation#9]
Hi @urialon!

I read the paper and have some questions and would be really grateful if you could answer. You mention your model generates likely completions given a piece of code with missing holes. Can your proposed model be used when the hole appears not in between some source code lines but at the end (like an autocomplete model)?

In your evaluation do you compute ASTs for complete code and then mask subtrees and their corresponding tokens for prediction in your experiments? or are there ways to compute ASTs for incomplete code? A previous work https://arxiv.org/pdf/2005.08025.pdf (section 5) mentions that ASTs can only be retrieved on complete code snippets that are syntactically correct, which is often not available for a code completion system. I wanted to understand how the setup you study is different from this one.

Thanks!

**Comment by urialon** [comment]
Hi @akhileshgotmare ,
Thank you for your interest in SLM!

Yes, the model can be used when there is only context "before" the completion" and not "after" the completion.
It may perform a little worse, though.

Regarding your second question - yes, we parse (compute AST) for complete code snippets, and then mask the subtree that we wish to generate.

In the backend of our demo website ( https://anycodegen.org/ ), we replace the question mark ("?" - the "hole") with some constant identifier like `HOLE`, which allows the parser to parse it as if it was a regular identifier, and then we traverse the AST to find this `HOLE` identifier, and mark it as the root of the subtree that we wish to generate.

So, to summarize, I agree that it's easier to parse complete code rather than partial code. 
At training time we avoid this because we first parse, and then mask the holes (or "cut" them from the tree).
At inference time (for example, the demo website), we add a dummy node instead of the question mark, so the code will parse.

I hope it helps.

Best,
Uri

**Comment by urialon** [comment]
Closing due to inactivity, feel free to re-open if you have any additional questions!
